### ISSUE TITLE
XRootD fails to start without reverse DNS entry

### ISSUE THREAD

**Issue by bbockelm** [xrootd/xrootd#2159]
If you don't have a reverse DNS on a local interface for your current hostname, then `xrootd` will fail to start with this relatively mysterious error message:

```
XrdConfig: Unable to determine host name;  ai_family not supported ; execution terminated.
```

To reproduce, simply do the following:

```
$ sudo unshare -u /bin/sh
# hostname
realhost.unl.edu
# hostname test.example.com
# xrootd -R xrootd
240101 14:24:56 1663941 XrdConfig: Unable to determine host name;  ai_family not supported ; execution terminated.
```

(a less easy way to reproduce is use GitHub's current Mac OS X runner which appears to only have a reverse DNS entry about 10% of the time)

It appears this is because:
- The determination of the process's hostname fails [here](https://github.com/xrootd/xrootd/blob/master/src/XrdNet/XrdNetIdentity.cc#L45), 
- `XrdConfig.cc` will create a `XrdAddrInfo` object using the default initializer, which should default to the process's hostname.
- The socket constructor notices the failure but proceeds with a fake address for the hostname lookup [here](https://github.com/xrootd/xrootd/blob/master/src/XrdNet/XrdNetAddr.cc#L101-L102).
- `Set` is called with an invalid hostname.  This will zero out the socket address, leaving it invalid.  A failure is returned but the failure is ignored.
- Later on, `XrdConfig.cc` will call `XrdAddrInfo::Name()` to get the hostname of the socket.  The `Resolve` method is called to do a reverse-DNS lookup on the socket address.  However, the socket address was previously memset to 0's, meaning [this conditional block](https://github.com/xrootd/xrootd/blob/master/src/XrdNet/XrdNetAddrInfo.cc#L384-L390) causes a return error code of `EAI_FAMILY`.
- The `EAI_FAMILY` code is converted to the mysterious `ai_family not supported` error message.

Now, I think all would agree that the error message just stinks.  There's a few options on how to better determine when there are no reverse DNS lookups available.

However, I also question whether this should be a hard failure.  Reverse DNS is mostly needed for setting up clustered mode with cmsd.  If you're using xrootd standalone (or, I suspect, as part of EOS), I don't think this should be necessary.  I can work around it in a GitHub runner as I've got sudo access and can muck around with `/etc/hosts`; that trick leaves out unprivileged users though!

If you're using virtualized environments (esp. Kubernetes), an IPv6-only setup (where PTR records are rarer), or just don't want to setup PTR records, I think XRootD should Just Work.

**Comment by abh3** [comment]
Well, as all of these things, it's complicated. If you look in the code you will see that, at least in XrdConfig.cc, we do not require reverse lookup or even DNS registry. We do, however, issue explicit messages saying it will likely cause various feature failures. Looking back, changes over the years involving IPv6 support as well as identity resolution, specifically for k8s environments, introduced processing errors when reverse lookup was not available. So, it's a matter of understanding what has to change to not introduce even more unintended interactions. Thanks for bringing this up. I suppose it shows that practically all sites implement reverse lookup. Anyway, a fix will be on its way.

**Comment by bbockelm** [comment]
> If you look in the code you will see that, at least in XrdConfig.cc, we do not require reverse lookup or even DNS registry.

I'm not quite following this comment.  When the resolution fails, `_exit(16)` is called: https://github.com/xrootd/xrootd/blob/master/src/Xrd/XrdConfig.cc#L620

Or are you just saying this was not the intention?

**Comment by abh3** [comment]
Yes, that's what I meant. This was an unintended feature interaction.
________________________________

**Comment by abh3** [comment]
OK, I have analyzed this and I don't see why this fails at all. I agree that all the reverse translation may not work but in the end if none of it works it will use the raw value that gethostname() returned. The comments indicate that gethostname() must succeed and if it does not then everything will fail. So, are we really saying that if there is no hostname we should not fail? More explicitly, we should simply use the IP address as the hostname. I say this because the only way the code at https://github.com/xrootd/xrootd/blob/master/src/XrdNet/XrdNetIdentity.cc#L45 fails is if gethostname() fails. Otherwise, the FQN value is gauranteed to not be zero. I have a feeling that another problem exists here, Since it appears you have a host that exhibits this problem (I do not even when reverse translation is unavailable) could you simply step through getMyFQN() at startup to see where and why it actually failed?

**Comment by abh3** [comment]
Correct. While we did not fix the horrible error message, we did make it moot. The identity of the node will be based on the last known representable identity. That may actually be nothing more than the IP address. However, as long as we have a unique identity for the host, startup will not fail. Oddly, this means that if the machine is not accessible beyond itself and there is no DNS entry for it, startup will fail.

**Comment by bbockelm** [comment]
Perhaps your host still does forward-resolution without networking (this would work if your OS uses systemd's resolvd; RHEL does not).

Try this:

```
$ sudo unshare -u -n /bin/sh -c 'getent hosts `hostname`'
$ echo $?
2
$ getent hosts `hostname`
2600:900:6:1301:5054:ff:fe0b:9cba hcc-briantest7.unl.edu
$ echo $?
0
```

That shows I've broken resolution with the `unshare` whereas it works outside.

You'd need some way to break forward resolution in this case (perhaps carefully turn off the relevant systemd component or mess with nssswitch.conf?).

**Comment by amadio** [comment]
Closing and marking as fixed in 5.7.1. Please reopen if the fix above is not sufficient. Thank you.
